File: go/src/infra/appengine/weetbix/frontend/handlers/xsrf.go
// Copyright 2022 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package handlers

import (
	"net/http"

	"go.chromium.org/luci/common/logging"
	"go.chromium.org/luci/server/auth/xsrf"
	"go.chromium.org/luci/server/router"
)

// GetXSRFToken serves a GET request for
// /api/xsrfToken.
func (h *Handlers) GetXSRFToken(ctx *router.Context) {
	tok, err := xsrf.Token(ctx.Context)
	if err != nil {
		logging.Errorf(ctx.Context, "Obtain XSRF token: %s", err)
		http.Error(ctx.Writer, "Internal server error.", http.StatusInternalServerError)
		return
	}

	data := map[string]string{
		"token": tok,
	}
	respondWithJSON(ctx, data)
}

File: go/src/infra/appengine/weetbix/frontend/ui/src/libs/xsrf.ts
// Copyright 2022 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

let xsrfToken : XSRFTokenCache | null = null;

// obtainXSRFToken obtains a current XSRF token, for interacting
// with Weetbix APIs.
export async function obtainXSRFToken(): Promise<string> {
    if (xsrfToken == null) {
        // Initialise with an expired token.
        xsrfToken = {
            token: "",
            expiry: new Date(0),
        }
    }
    const now = new Date()
    if (now <= xsrfToken.expiry) {
        // Token is still valid.
        return xsrfToken.token;
    }

    // Refresh the token.
    const r = await fetch(`/api/xsrfToken`);
    const response : XSRFToken = await r.json();

    // Let the token expire 3 hours from now. Actual validity
    // is 4 hours, so this leaves some margin.
    const expiry = new Date()
    expiry.setTime(expiry.getTime() + 3*60*60*1000)

    xsrfToken = {
        token: response.token,
        expiry: expiry,
    }

    return response.token
}

// An XSRF token as returned by the server.
interface XSRFToken {
    token: string;
}

interface XSRFTokenCache {
    token: string;
    expiry: Date;
}
